satellite/satellitedb/containment.go

```diff
@@ -0,0 +1,97 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package satellitedb
+
+import (
+	"context"
+	"database/sql"
+
+	"storj.io/storj/pkg/audit"
+	"storj.io/storj/pkg/pb"
+	"storj.io/storj/pkg/storj"
+	dbx "storj.io/storj/satellite/satellitedb/dbx"
+)
+
+type containment struct {
+	db *dbx.DB
+}
+
+// Get gets the pending audit by node id
+func (containment *containment) Get(ctx context.Context, id pb.NodeID) (*audit.PendingAudit, error) {
+	if id.IsZero() {
+		return nil, audit.ContainError.New("node ID empty")
+	}
+
+	pending, err := containment.db.Get_PendingAudits_By_NodeId(ctx, dbx.PendingAudits_NodeId(id.Bytes()))
+	if err == sql.ErrNoRows {
+		return nil, audit.ErrContainedNotFound.New(id.String())
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	return convertDBPending(pending)
+}
+
+// IncrementPending will create a new pending audit entry, or increase its reverify count if it already exists
+func (containment *containment) IncrementPending(ctx context.Context, pendingAudit *audit.PendingAudit) error {
+	statement := containment.db.Rebind(
+		`INSERT INTO pending_audits (node_id, piece_id, stripe_index, share_size, expected_share_hash, reverify_count)
+		VALUES (?, ?, ?, ?, ?, ?)
+		ON CONFLICT(node_id)
+		DO UPDATE SET reverify_count = pending_audits.reverify_count + 1`,
+	)
+	_, err := containment.db.ExecContext(ctx, statement,
+		pendingAudit.NodeID.Bytes(), pendingAudit.PieceID.Bytes(), pendingAudit.StripeIndex, pendingAudit.ShareSize, pendingAudit.ExpectedShareHash, pendingAudit.ReverifyCount,
+	)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// Delete deletes the pending audit
+func (containment *containment) Delete(ctx context.Context, id pb.NodeID) error {
+	if id.IsZero() {
+		return audit.ContainError.New("node ID empty")
+	}
+	isDeleted, err := containment.db.Delete_PendingAudits_By_NodeId(ctx, dbx.PendingAudits_NodeId(id.Bytes()))
+	if err == sql.ErrNoRows {
+		return audit.ErrContainedNotFound.New(id.String())
+	}
+	if !isDeleted {
+		return audit.ErrContainDelete.New(id.String(), err)
+	}
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func convertDBPending(info *dbx.PendingAudits) (*audit.PendingAudit, error) {
+	if info == nil {
+		return nil, Error.New("missing info")
+	}
+
+	nodeID, err := storj.NodeIDFromBytes(info.NodeId)
+	if err != nil {
+		return nil, err
```

Comment:
It's commented above as a reply of the @VinozzZ comment, but here it's were I see clear that we  can loose more lines in the stack trace than one

This apply for any non-wrapped error returned by this function

Reply:
okay so I should use errs.Wrap(err) here? or ContainError.Wrap here instead of at the layer up?

Reply:
One of those should be good for the stack trace matter.

Which one, it's less obvious to me and somebody could have a strong argument on which one based on how it has been used so far in other parts of the code base.

I'm torn which one should be, from one part I see that using `errs.Wrap` makes the function not to hold any responsibility about the who are the callers but at the same time it's an unexported function so the callers are from only from this package, so using `ContainError.Wrap` allow the callers to directly return from it, however it less flexible to future callers to use it if they have to return a different error class (e.g. `ErrContainedNotFound`).

I would say to seek advice from some one with more experience in the Storj code base for that answer.

Reply:
FWIW, there is no stack trace lost here, because `storj.NodeIDFromBytes` already return `zeebo/errs` error.

Reply:
yes, that's right.
My bad, I didn't see that it was a Storj function so I missed to check if it was returning.